cmd/server/storage/db/dml/repository.go

```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		var pgconnErr *pgconn.PgError
+		if errors.As(err, &pgconnErr) && pgerrcode.IsConnectionException(pgconnErr.Code) {
+			result, err = retryExecContext(dbr, internalErrors.TriesTimes, ctx, query, args)
+			if err == nil {
+				return result, nil
+			}
+		}
+		return nil, internalErrors.NewDBError(err)
+	}
+	return result, nil
+}
+
+func (dbr *Repository) QueryRowContext(ctx context.Context, query string, args ...any) *sql.Row {
+	row := dbr.Database.QueryRowContext(ctx, query, args...)
+	if row.Err() != nil {
+		var pgconnErr *pgconn.PgError
+		if errors.As(row.Err(), &pgconnErr) && pgerrcode.IsConnectionException(pgconnErr.Code) {
+			row = retryQueryRowContext(dbr, internalErrors.TriesTimes, ctx, query, args)
+		}
+	}
+	return row
+}
+
+func (dbr *Repository) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	rows, err := dbr.Database.QueryContext(ctx, query, args...)
+
+	if err != nil {
+		var pgconnErr *pgconn.PgError
+		if errors.As(err, &pgconnErr) && pgerrcode.IsConnectionException(pgconnErr.Code) {
+			rows, err = retryQueryContext(dbr, internalErrors.TriesTimes, ctx, query, args)
+			if err != nil {
+				return nil, internalErrors.NewDBError(err)
+			}
+		}
+	}
+	return rows, nil
+}
+
+func (dbr *Repository) Close() error {
+	err := dbr.Database.Close()
+	if err != nil {
+		return internalErrors.NewDBError(err)
+	}
+	return nil
+}
+
+func (dbr *Repository) PingContext(ctx context.Context) error {
+	err := dbr.Database.PingContext(ctx)
+	if err != nil {
+		var pgconnErr *pgconn.PgError
+		if errors.As(err, &pgconnErr) && pgerrcode.IsConnectionException(pgconnErr.Code) {
+			err = retryPing(dbr, ctx, internalErrors.TriesTimes)
+			if err != nil {
+				return internalErrors.NewDBError(err)
+			}
+		}
+	}
+	return err
+}
+
+func (dbr *Repository) SetDatabase() error {
+	database, err := dbr.Open()
+	if err != nil {
+		logger.WriteErrorLog("Database open error", err.Error())
+		return internalErrors.NewDBError(err)
+	}
+	dbr.Database = database
+	return nil
+}
+
+func (dbr *Repository) Open() (*sql.DB, error) {
+	result, err := sql.Open("pgx", handlers.DatabaseDSN)
+	if err != nil {
+		var pgconnErr *pgconn.PgError
+		if errors.As(err, &pgconnErr) && pgerrcode.IsConnectionException(pgconnErr.Code) {
+			result, err = retryOpen("pgx", handlers.DatabaseDSN, internalErrors.TriesTimes)
+			if err != nil {
+				return nil, internalErrors.NewDBError(err)
+			}
+		}
+	}
+	return result, nil
+}
+
+func NewRepository() *Repository {
+	rep := &Repository{}
+	_ = rep.SetDatabase()
+	return rep
+}
+
+func CreateOrUpdateCounter(dbr *Repository, name string, value models.Counter) (sql.Result, error) {
+	var err error
+	row := dbr.QueryRowContext(context.Background(), "SELECT name FROM counter WHERE name = $1", name)
+
+	if row.Err() != nil {
+		logger.WriteErrorLog("AddCounter database query error", row.Err().Error())
+	}
+	var selectedName string
+	_ = row.Scan(&selectedName)
```

Comment:
```suggestion
	err = row.Scan(&selectedName)
```
О том, что строка не найдена можно узнать ТОЛЬКО из анализа типа ошибки. Потому что отсутствие соединения с БД точно так же вернет пустую selectedName.

А вообще в постгресе для апсертов есть прекрасная конструкция:
```
INSERT INTO counter (name, value) VALUES ($1, $2)
ON CONFLICT (name) SET value = value + $2
```
Только исключительный индекс надо на поле name поставить.